File: 2691-count-vowel-strings-in-ranges/count-vowel-strings-in-ranges.py
class Solution(object):
    def vowelStrings(self, words, queries):
        """
        :type words: List[str]
        :type queries: List[List[int]]
        :rtype: List[int]
        """
        vowels = {'a', 'e', 'i', 'o', 'u'}

        presum = [0]
        for word in words:
            presum.append(presum[-1] + (1 if word[0] in vowels and word[-1] in vowels else 0))
        
        return [presum[end + 1] - presum[start] for start, end in queries]

        

File: 2691-count-vowel-strings-in-ranges/count-vowel-strings-in-ranges.java
class Solution {
    public int[] vowelStrings(String[] words, int[][] queries) {
        int n = words.length;
        int m = queries.length;
        int[] presum =  new int[n + 1];
        int[] ans =  new int[m];


        if(isStartEndVowel(words[0])) presum[0] = 1;

        for(int i = 1; i < n; ++i) {
            if(isStartEndVowel(words[i])) {
                presum[i] = presum[i - 1] + 1;
            }else {
                presum[i] = presum[i - 1];
            }
        }


        for(int i = 0; i < m; ++i) {
            int start = queries[i][0];
            int end = queries[i][1];
            
            ans[i] = presum[end];
            if(start - 1 >= 0) 
                ans[i] -= presum[start - 1];
        }
        return ans;
    }

    private boolean isStartEndVowel(String s){
        char start = s.charAt(0);
        char end = s.charAt(s.length() - 1);
        return (start == 'a' ||  start == 'e' || start == 'i' || start == 'o' || start == 'u')
            && (end == 'a' || end == 'e' || end == 'i' ||  end == 'o' || end == 'u');
    }
}